src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/LanguageClient/Cohost/CohostSemanticTokensRangeEndpoint.cs

```diff
@@ -66,19 +70,43 @@ internal sealed class CohostSemanticTokensRangeEndpoint(
 
     protected override async Task<SemanticTokens?> HandleRequestAsync(SemanticTokensRangeParams request, RazorCohostRequestContext context, CancellationToken cancellationToken)
     {
-        var correlationId = Guid.NewGuid();
-        using var _ = _telemetryReporter.TrackLspRequest(Methods.TextDocumentSemanticTokensRangeName, RazorLSPConstants.CohostLanguageServerName, correlationId);
+        var razorDocument = context.TextDocument.AssumeNotNull();
+        var span = request.Range.ToLinePositionSpan();
 
-        var data = await _semanticTokensInfoService.GetSemanticTokensDataAsync(context.TextDocument.AssumeNotNull(), request.Range.ToLinePositionSpan(), correlationId, cancellationToken);
+        var remoteClient = await _remoteClientProvider.TryGetClientAsync(cancellationToken).ConfigureAwait(false);
 
-        if (data is null)
+        if (remoteClient is null)
         {
+            _logger.LogWarning($"Couldn't get remote client");
             return null;
         }
 
-        return new SemanticTokens
+        try
         {
-            Data = data
-        };
+            var colorBackground = _clientSettingsManager.GetClientSettings().AdvancedSettings.ColorBackground;
+
+            var correlationId = Guid.NewGuid();
+            using var _ = _telemetryReporter.TrackLspRequest(Methods.TextDocumentSemanticTokensRangeName, RazorLSPConstants.CohostLanguageServerName, correlationId);
+
+            var data = await remoteClient.TryInvokeAsync<IRemoteSemanticTokensService, int[]?>(
+                razorDocument.Project.Solution,
+                (service, solutionInfo, cancellationToken) => service.GetSemanticTokensDataAsync(solutionInfo, razorDocument.Id, span, colorBackground, correlationId, cancellationToken),
+                cancellationToken).ConfigureAwait(false);
+
+            if (data.Value is not { } tokens)
+            {
+                return null;
+            }
+
+            return new SemanticTokens
+            {
+                Data = tokens
+            };
+        }
+        catch (Exception ex)
+        {
+            _logger.LogError(ex, $"Error calling remote");
```

Comment:
Should this also go to telemetry?

Reply:
Good point. We do get whatever Roslyn has for free (ie, we get yellow bars for razor services etc.), but can't hurt logging our own too. Will leave as a follow up though, am planning on simplifiying the `IRemoteClientProvider` API tonight in the car, as everything current copies around the same boilerplate, and its annoying.